docs/proposals/20210335-load-balancer-provider.md

```diff
@@ -0,0 +1,999 @@
+- Add optional LoadBalancerRef (if needed, based on the provider)
+
+#### New Watches:
+
+- LoadBalancer
+
+##### Example: KubeadmControlPlane
+
+```yaml
+apiVersion: controlplane.cluster.x-k8s.io/v1alpha4
+kind: KubeadmControlPlane
+metadata:
+  name: my-cluster-control-plane
+spec:
+  endpoint:
+    host: test.example.com
+    port: 443
+  loadBalancerRef:
+    apiGroup: cluster.x-k8s.io
+    kind: LoadBalancer
+    name: my-cluster-control-plane-lb
+```
+
+#### Changes to ControlPlane provider reconciliation:
+
+##### Example: KubeadmControlPlane
+
+![Kubeadm Control Plane Reconciliation Flow](images/load-balancer/kcp-reconciliation.png)
+
+###### Scenario: Load Balancer specified, externally managed endpoint
+
+User wants to provide an externally managed endpoint for use when creating a cluster, but wants Cluster API to manage the load balancer for the control plane.
+
+Preconditions:
+- `Spec.Endpoint` is non-empty (could be populated by the Cluster controller)
+- `Spec.LoadBalancerRef` is defined
+
+Reconciliation:
+TODO
+
+###### Scenario: Load Balancer specified, provider managed endpoint
+
+User wants Cluster API to manage the endpoint for the control plane when creating a cluster.
+
+Preconditions:
+- `Spec.Endpoint` is empty
+- `Spec.LoadBalancerRef` is defined
+
+Reconciliation:
+TODO
+
+###### Scenario: Load Balancer not specified (fallback behavior)
+
+User wants Cluster API to manage the endpoint for the control plane, but is using a provider that doesn't yet support pluggable load balancers.
+
+Preconditions:
+- `Spec.Endpoint` is empty
+- `Spec.LoadBalancerRef` is nil
+
+Reconciliation:
+TODO
+
+### New Types
+
+#### Load Balancer
+
+A common type to aggregate information between Load Balancer provider implementations and to provide some common reconciliation logic.
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1alpha4
+kind: LoadBalancer
+  metadata:
+    name: my-cluster-control-plane-lb
+spec:
+  clusterName: my-cluster
+  endpoint:
+    host: test.example.com
+    port: 443
+  targetKind:
+    apiGroup: cluster.x-k8s.io
+    kind: Machine
+  selector:
+    matchExpressions:
+    matchLabels:
+      cluster.x-k8s.io/cluster-name: my-cluster
+      loadbalancer.cluster.x-k8s.io/name: my-cluster-control-plane-lb
+  infrastructureRef:
+    apiGroup: infrastructure.cluster.x-k8s.io
+    kind: AWSNLB
+    name: my-cluster-control-plane-nlb
+status:
+  targetKind: <string representing group/kind name for selector>
+  targetSelector: <string representing the label selector for the referenced LB>
+  targets: (List of members matching label selector)
+  activeTargets: (List of active members)
+  conditions: (List of infra providers condition for the load balancer)
+```
+
+#### Annotations
+
+- Used by providers to signal the state of the resource in the LB
+  - loadbalancer.cluster.x-k8s.io/name: `<name of lb>`
+  - loadbalancer.cluster.x-k8s.io/status: `<status>`
```

Comment:
Are infrastructure providers in charge of setting this annotation on a LoadBalancer resource? If so, wouldn't it be more useful to have these values under spec or status instead and have them as a contract?